File: mobile/android/OfflineSyncSample.kt
package com.assetra.sync

import java.time.Instant
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

data class LocalScanEvent(
    val clientEventId: String,
    val symbology: String,
    val rawValue: String,
    val sourceType: String,
    val capturedAt: String,
    val synced: Boolean,
)

interface SyncApi {
    suspend fun pushPull(
        lastSyncAt: String?,
        scanEvents: List<LocalScanEvent>,
        conflictAcknowledgements: List<ConflictAcknowledgement>,
    ): SyncResponse
}

data class SyncResponse(
    val serverTime: String,
    val acceptedScanEventIds: List<Int>,
    val assetChanges: List<Map<String, Any?>>,
    val conflicts: List<ConflictRecord>,
    val conflictStrategy: String,
)

class OfflineSyncEngine(
    private val localStore: LocalStore,
    private val syncApi: SyncApi,
) {
    suspend fun sync() = withContext(Dispatchers.IO) {
        val pending = localStore.pendingScanEvents()
        val pendingAcks = localStore.pendingConflictAcks()
        val response = syncApi.pushPull(localStore.lastSyncAt(), pending, pendingAcks)
        localStore.markSynced(pending.map { it.clientEventId })
        localStore.markConflictAcksSynced(pendingAcks.map { it.conflictId })
        localStore.upsertAssets(response.assetChanges)
        localStore.saveConflicts(response.conflicts)
        localStore.saveLastSyncAt(response.serverTime)
    }
}

interface LocalStore {
    suspend fun pendingScanEvents(): List<LocalScanEvent>
    suspend fun markSynced(clientEventIds: List<String>)
    suspend fun upsertAssets(assets: List<Map<String, Any?>>)
    suspend fun saveConflicts(conflicts: List<ConflictRecord>)
    suspend fun pendingConflicts(): List<ConflictRecord>
    suspend fun resolveConflict(conflictId: String, useServerValue: Boolean)
    suspend fun pendingConflictAcks(): List<ConflictAcknowledgement>
    suspend fun markConflictAcksSynced(conflictIds: List<String>)
    suspend fun saveLastSyncAt(serverTime: String)
    suspend fun lastSyncAt(): String?
}

class InMemoryLocalStore : LocalStore {
    private val events = mutableListOf<LocalScanEvent>()
    private val assets = mutableMapOf<String, Map<String, Any?>>()
    private val conflicts = mutableListOf<ConflictRecord>()
    private val conflictAcks = mutableListOf<ConflictAcknowledgement>()
    private var lastSync: String? = null

    override suspend fun pendingScanEvents(): List<LocalScanEvent> {
        return events.filter { !it.synced }
    }

    override suspend fun markSynced(clientEventIds: List<String>) {
        val updated = events.map { event ->
            if (clientEventIds.contains(event.clientEventId)) {
                event.copy(synced = true)
            } else {
                event
            }
        }
        events.clear()
        events.addAll(updated)
    }

    override suspend fun upsertAssets(assets: List<Map<String, Any?>>) {
        for (asset in assets) {
            val id = asset["id"]?.toString() ?: continue
            this.assets[id] = asset
        }
    }

    override suspend fun saveConflicts(conflicts: List<ConflictRecord>) {
        this.conflicts.addAll(conflicts)
    }

    override suspend fun pendingConflicts(): List<ConflictRecord> = conflicts.toList()

    override suspend fun resolveConflict(conflictId: String, useServerValue: Boolean) {
        val conflict = conflicts.firstOrNull { it.id == conflictId }
        if (!useServerValue && conflict != null) {
            val replayEvent = LocalScanEvent(
                clientEventId = UUID.randomUUID().toString(),
                symbology = "conflict_replay",
                rawValue = "asset=${conflict.assetId};field=${conflict.field};value=${conflict.localValue}",
                sourceType = "conflict_replay",
                capturedAt = Instant.now().toString(),
                synced = false,
            )
            events.add(replayEvent)

            val current = assets[conflict.assetId]?.toMutableMap() ?: mutableMapOf("id" to conflict.assetId)
            current[conflict.field] = conflict.localValue
            assets[conflict.assetId] = current
        }

        conflicts.removeAll { it.id == conflictId }
        conflictAcks.add(
            ConflictAcknowledgement(
                conflictId = conflictId,
                resolution = if (useServerValue) "accept_server" else "keep_local",
                resolvedAt = Instant.now().toString(),
            )
        )
    }

    override suspend fun pendingConflictAcks(): List<ConflictAcknowledgement> = conflictAcks.toList()

    override suspend fun markConflictAcksSynced(conflictIds: List<String>) {
        conflictAcks.removeAll { conflictIds.contains(it.conflictId) }
    }

    override suspend fun saveLastSyncAt(serverTime: String) {
        lastSync = serverTime
    }

    override suspend fun lastSyncAt(): String? = lastSync

    // Helper for demos to add local events.
    fun addLocalScanEvent(event: LocalScanEvent) {
        events.add(event)
    }

    // Helper for demos to seed conflict records.
    fun seedConflicts(seed: List<ConflictRecord>) {
        conflicts.clear()
        conflicts.addAll(seed)
    }
}

fun sampleCapture(localStore: InMemoryLocalStore, rawValue: String) {
    val event = LocalScanEvent(
        clientEventId = UUID.randomUUID().toString(),
        symbology = "qr",
        rawValue = rawValue,
        sourceType = "camera",
        capturedAt = Instant.now().toString(),
        synced = false,
    )
    localStore.addLocalScanEvent(event)
}

// Example usage (wire into your app lifecycle).
//
// val tokenProvider = object : AccessTokenProvider {
//     override fun accessToken(): String = currentAccessToken
//     override fun refreshAccessToken(): String {
//         val refreshed = authClient.refresh(currentRefreshToken)
//         currentAccessToken = refreshed.access
//         currentRefreshToken = refreshed.refresh
//         return currentAccessToken
//     }
// }
// val api = AssetraApi(
//     baseUrl = "https://api.assetra.example",
//     tenantId = "1",
//     tokenProvider = tokenProvider,
// )
// val localStore = InMemoryLocalStore()
// sampleCapture(localStore, rawValue = "QR-100")
// val engine = OfflineSyncEngine(localStore, api)
// engine.sync()

File: mobile/ios/OfflineSyncSample.swift
import Foundation

struct LocalScanEvent: Codable {
    let clientEventId: String
    let symbology: String
    let rawValue: String
    let sourceType: String
    let capturedAt: String
    let synced: Bool
}

struct SyncResponse: Codable {
    let serverTime: String
    let acceptedScanEventIds: [Int]
    let assetChanges: [[String: JSONValue]]
    let conflicts: [ConflictRecord]
    let conflictStrategy: String
}

protocol SyncAPI {
    func pushPull(
        lastSyncAt: String?,
        scanEvents: [LocalScanEvent],
        conflictAcknowledgements: [ConflictAcknowledgement]
    ) async throws -> SyncResponse
}

protocol LocalStore {
    func pendingScanEvents() async -> [LocalScanEvent]
    func markSynced(clientEventIds: [String]) async
    func upsertAssets(_ assets: [[String: JSONValue]]) async
    func saveConflicts(_ conflicts: [ConflictRecord]) async
    func pendingConflicts() async -> [ConflictRecord]
    func resolveConflict(conflictId: String, useServerValue: Bool) async
    func pendingConflictAcks() async -> [ConflictAcknowledgement]
    func markConflictAcksSynced(conflictIds: [String]) async
    func saveLastSyncAt(_ value: String) async
    func lastSyncAt() async -> String?
}

final class InMemoryLocalStore: LocalStore {
    private var events: [LocalScanEvent] = []
    private var assets: [String: [String: JSONValue]] = [:]
    private var conflicts: [ConflictRecord] = []
    private var conflictAcks: [ConflictAcknowledgement] = []
    private var lastSync: String?

    func pendingScanEvents() async -> [LocalScanEvent] {
        events.filter { !$0.synced }
    }

    func markSynced(clientEventIds: [String]) async {
        events = events.map { event in
            if clientEventIds.contains(event.clientEventId) {
                return LocalScanEvent(
                    clientEventId: event.clientEventId,
                    symbology: event.symbology,
                    rawValue: event.rawValue,
                    sourceType: event.sourceType,
                    capturedAt: event.capturedAt,
                    synced: true
                )
            }
            return event
        }
    }

    func upsertAssets(_ assets: [[String: JSONValue]]) async {
        for asset in assets {
            if case let .string(id)? = asset["id"] {
                self.assets[id] = asset
            } else if case let .number(id)? = asset["id"] {
                self.assets[String(id)] = asset
            }
        }
    }

    func saveConflicts(_ conflicts: [ConflictRecord]) async {
        self.conflicts.append(contentsOf: conflicts)
    }

    func pendingConflicts() async -> [ConflictRecord] {
        conflicts
    }

    func resolveConflict(conflictId: String, useServerValue: Bool) async {
        if !useServerValue, let conflict = conflicts.first(where: { $0.id == conflictId }) {
            let replayEvent = LocalScanEvent(
                clientEventId: UUID().uuidString,
                symbology: "conflict_replay",
                rawValue: "asset=\(conflict.assetId);field=\(conflict.field);value=\(conflict.localValue)",
                sourceType: "conflict_replay",
                capturedAt: ISO8601DateFormatter().string(from: Date()),
                synced: false
            )
            events.append(replayEvent)

            var current = assets[conflict.assetId] ?? ["id": .string(conflict.assetId)]
            current[conflict.field] = .string(conflict.localValue)
            assets[conflict.assetId] = current
        }

        conflicts.removeAll { $0.id == conflictId }
        conflictAcks.append(
            ConflictAcknowledgement(
                conflictId: conflictId,
                resolution: useServerValue ? "accept_server" : "keep_local",
                resolvedAt: ISO8601DateFormatter().string(from: Date())
            )
        )
    }

    func pendingConflictAcks() async -> [ConflictAcknowledgement] {
        conflictAcks
    }

    func markConflictAcksSynced(conflictIds: [String]) async {
        conflictAcks.removeAll { conflictIds.contains($0.conflictId) }
    }

    func saveLastSyncAt(_ value: String) async {
        lastSync = value
    }

    func lastSyncAt() async -> String? {
        lastSync
    }

    // Helper for demos to add local events.
    func addLocalScanEvent(_ event: LocalScanEvent) {
        events.append(event)
    }

    // Helper for demos to seed conflict records.
    func seedConflicts(_ seed: [ConflictRecord]) {
        conflicts = seed
    }
}

func sampleCapture(localStore: InMemoryLocalStore, rawValue: String) {
    let formatter = ISO8601DateFormatter()
    let event = LocalScanEvent(
        clientEventId: UUID().uuidString,
        symbology: "qr",
        rawValue: rawValue,
        sourceType: "camera",
        capturedAt: formatter.string(from: Date()),
        synced: false
    )
    localStore.addLocalScanEvent(event)
}

final class OfflineSyncEngine {
    private let store: LocalStore
    private let api: SyncAPI

    init(store: LocalStore, api: SyncAPI) {
        self.store = store
        self.api = api
    }

    func sync() async throws {
        let pending = await store.pendingScanEvents()
        let pendingAcks = await store.pendingConflictAcks()
        let response = try await api.pushPull(
            lastSyncAt: await store.lastSyncAt(),
            scanEvents: pending,
            conflictAcknowledgements: pendingAcks
        )
        await store.markSynced(clientEventIds: pending.map { $0.clientEventId })
        await store.markConflictAcksSynced(conflictIds: pendingAcks.map { $0.conflictId })
        await store.upsertAssets(response.assetChanges)
        await store.saveConflicts(response.conflicts)
        await store.saveLastSyncAt(response.serverTime)
    }
}

// Example usage (wire into your app lifecycle).
//
// struct TokenProvider: AccessTokenProvider {
//     var access: String
//     var refresh: String
//     let authClient: AuthClient
//
//     func accessToken() -> String { access }
//
//     func refreshAccessToken() async throws -> String {
//         let tokens = try await authClient.refresh(refreshToken: refresh)
//         // Persist tokens in your store as needed.
//         return tokens.access
//     }
// }
// let api = AssetraAPI(
//     baseURL: URL(string: "https://api.assetra.example")!,
//     tenantId: "1",
//     tokenProvider: tokenProvider
// )
// let localStore = InMemoryLocalStore()
// sampleCapture(localStore: localStore, rawValue: "QR-100")
// let engine = OfflineSyncEngine(store: localStore, api: api)
// Task { try await engine.sync() }
